lib/charms/opensearch/v0/opensearch_plugins.py

```diff
@@ -0,0 +1,348 @@
+        ...
+
+        self.my_plugin = MyPlugin(self)
+
+    ...
+
+    def _on_update_status(self, event):
+        ...
+
+        # Check my plugin status
+        if self.model.get_relation("my-plugin-relation") is not None:
+            self.unit.status = self.my_plugin.status()
+
+        for relation in self.model.relations.get(ClientRelationName, []):
+            self.opensearch_provider.update_endpoints(relation)
+
+        self.user_manager.remove_users_and_roles()
+        # If relation not broken - leave
+        if self.model.get_relation("certificates") is not None:
+            return
+        # handle when/if certificates are expired
+        self._check_certs_expiration(event)
+"""
+
+import logging
+import os
+from abc import abstractmethod
+from typing import Any, Dict, List, Optional, Tuple
+
+from charms.opensearch.v0.helper_enums import BaseStrEnum
+from charms.opensearch.v0.opensearch_exceptions import OpenSearchError
+from jproperties import Properties
+
+# The unique Charmhub library identifier, never change it
+LIBID = "3b05456c6e304680b4af8e20dae246a2"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+logger = logging.getLogger(__name__)
+
+
+class OpenSearchPluginError(OpenSearchError):
+    """Exception thrown when an opensearch plugin is invalid."""
+
+
+class OpenSearchPluginMissingDepsError(OpenSearchPluginError):
+    """Exception thrown when an opensearch plugin misses installed dependencies."""
+
+    def __init__(self, name: str, missing_dependencies: List[str]):
+        self._deps = missing_dependencies
+        self._name = name
+
+    def __str__(self):
+        """Converts exception to a string."""
+        return f"Failed to install {self._name}, missing dependencies: {self._deps}"
+
+
+class OpenSearchPluginInstallError(OpenSearchPluginError):
+    """Exception thrown when opensearch plugin installation fails."""
+
+    def __init__(self, name: str, msg: str):
+        self._msg = msg
+        self._name = name
+
+    def __str__(self):
+        """Converts exception to a string."""
+        return f"Failed to install plugin {self._name}: {self._msg}"
+
+
+class OpenSearchPluginRemoveError(OpenSearchPluginError):
+    """Exception thrown when opensearch plugin removal fails."""
+
+    def __init__(self, name: str, msg: str):
+        self._msg = msg
+        self._name = name
+
+    def __str__(self):
+        """Converts exception to a string."""
+        return f"Failed to remove plugin {self._name}: {self._msg}"
+
+
+class PluginState(BaseStrEnum):
+    """Enum for the states possible in plugins' lifecycle."""
+
+    MISSING = "missing"
+    AVAILABLE = "available"
+    ENABLED = "enabled"
+    WAITING_FOR_UPGRADE = "waiting-for-upgrade"
+
+
+class OpenSearchPluginConfig:
```

Comment:
see the comment above about potentially enriching `config_entries` and making a separate class for that